### PR TITLE
fix: memory issue when push large bentos

### PR DESCRIPTION
**Pull request by xianml**
## What does this PR address?

supporting limit max memory usage when pushing models

image

```
bentoml push facebook--opt-2.7b-service:905a4b602cda5c501f1b3a2650a4152680238254  --maxmemory 2
```

Test case 1: 
pushing `bento google--flan-t5-large-service`, model size 2.92 GiB

- no limit

1. time consumed: 3min 58s
2. memory usage: ~ 3GB

- maxmemory = 1

1. time consumed:4min 25s
2. memory usage: <1G

Test case 2:
pushing `bento google--flan-t5-large-service`, model size 12.55 GiB
- maxmemory = 3
image

1. time consumed:4min 48s
2. memory usage: max ~ 4G

Fixes #(issue)

## Before submitting:

- [x] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [x] Does the code follow BentoML's code style, `pre-commit run -a` script has passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [x] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

**Comment by jianshen92** [comment]
What are the bugs within the current implementation?

**Comment by xianml** [comment]
> What are the bugs within the current implementation?

context in https://bentoml-team.slack.com/archives/C02QLC8RB5W/p1695088745009929

as a summary, currently it will take too much memory when do a `bentoml push` since it will

**Comment by xianml** [comment]
> What are the bugs within the current implementation?

context in https://bentoml-team.slack.com/archives/C02QLC8RB5W/p1695088745009929

as a summary, currently it will take too much memory when do a `bentoml push` since it use io.BytesIO. So this fix is to use SpooledTemporaryFile instead to cap the memory usage

**Comment by sauyon** [comment]
Looks like unit tests are failing, maybe because of the requests change...?

Should we just use a `SpooledTemporaryFile` for everything?

**Comment by aarnphm** [comment]
@sauyon you need to change the tests patch `requests` to `httpx`. Let's open a separately PR to fix the test?

**Comment by xianml** [comment]
> Looks like unit tests are failing, maybe because of the requests change...?
> 
> Should we just use a `SpooledTemporaryFile` for everything?



> Looks like unit tests are failing, maybe because of the requests change...?
> 
> Should we just use a `SpooledTemporaryFile` for everything?

- checked the ut failed logs, seems our test case is out of date
- now, i just use SpooledTemporaryFile for pushing models.  Shall we replaced it one by one ? i am not very confident with a big code change.

File: src/bentoml_cli/bentos.py [hunk 1]
```diff
@@ -292,6 +298,7 @@ def push(shared_options: SharedOptions, bento_tag: str, force: bool, threads: in
             force=force,
             threads=threads,
             context=shared_options.cloud_context,
+            maxmemory=maxmemory,
```
Comment (aarnphm) [inline src/bentoml_cli/bentos.py:301]: %s/maxmemory/max_memory?

**Review by aarnphm — APPROVED** [review]
LGTM. @Haivilo worth to also test this once the deployment CLI/SDK is more mature and added some test case.

**Review by aarnphm — APPROVED** [review]

**Pull request closed without merging**